### PR TITLE
Add coding.md into docs

### PR DESCRIPTION
**Pull request by humblec**
Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

**Comment by humblec** [comment]
@gman0 @Madhu-1 @rootfs PTAL

File: README.md [hunk 1]
```diff
@@ -20,6 +20,12 @@ CephFS CSI plugins, see documentation in `docs/`.
 
 For example usage of RBD and CephFS CSI plugins, see examples in `examples/`.
 
+## Contributing to this repo
+
+Please follow [contibution guide](https://github.com/ceph/ceph-csi/tree/csi-v1.0/docs/coding.md)
```
Comment (Madhu-1) [inline README.md]: `csi-v1.0` or `master`?

File: README.md [hunk 2]
```diff
@@ -20,6 +20,12 @@ CephFS CSI plugins, see documentation in `docs/`.
...
+and [development-guide](https://github.com/ceph/ceph-csi/tree/csi-v1.0/docs/development-guide.md)
```
Comment (Madhu-1) [inline README.md]: same here

File: docs/development-guide.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+* Take the [A Tour of Go](http://tour.golang.org/welcome/1) course.
+* [Set up](https://golang.org/doc/code.html) Go development environment on your machine.
+* Read [Effective Go](https://golang.org/doc/effective_go.html) for best practices.
+
+## Development Workflow
+
+### Workspace and repository setup
+
+1. [Download](https://golang.org/dl/) Go (>=1.9) and [install](https://golang.org/doc/install) it on your system.
```
Comment (Madhu-1) [inline docs/development-guide.md]: go version need to be 1.11.x or greater

File: docs/development-guide.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+1. `$ git checkout master`
+2. `$ git pull`
+3. `$ git checkout -b hellopatch`
+
+Do your work here and commit.
+
+Run the test suite, which includes linting checks, static code check, and unit
+tests:
+
+`$ make tests`
```
Comment (Madhu-1) [inline docs/development-guide.md]: `make test`

File: README.md (hunk 2 above)
Comment (humblec) [reply README.md]: Done

File: docs/development-guide.md (hunk 3 above)
Comment (humblec) [reply docs/development-guide.md]: Done.

File: docs/development-guide.md (hunk 4 above)
Comment (humblec) [reply docs/development-guide.md]: Done

**Comment by humblec** [comment]
@Madhu-1 addressed the comments. Thanks!

File: docs/coding.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  Exception: defer()'d functions.
+
+### Imports
+
+We use the following convention for specifying imports:
+
+```
+<import standard library packages>
+
+<import ceph packages>
```
Comment (Madhu-1) [inline docs/coding.md]: `ceph`  to `ceph-csi`

**Comment by Madhu-1** [comment]
@humblec CI failure PTAL 

```
$ scripts/lint-text.sh --require-all
mdl --style scripts/mdl-style.rb ./README.md 
mdl --style scripts/mdl-style.rb ./docs/coding.md 
./docs/coding.md:13: MD001 Header levels should only increment by one level at a time
./docs/coding.md:45: MD010 Hard tabs
./docs/coding.md:46: MD010 Hard tabs
./docs/coding.md:13: MD022 Headers should be surrounded by blank lines
./docs/coding.md:14: MD032 Lists should be surrounded by blank lines
```

**Comment by humblec** [comment]
@Madhu-1 sure. will be on it soon.

**Comment by humblec** [comment]
>@humblec do we need to mention that we have different branches for csi-v1.0.0 and csi-v0.3.0
>
I think we can skip this for now.

File: docs/coding.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+Here's a list of some more specific conventions that are often followed in
+the code and will be pointed out in the review process:
+
+## General
+
+* Keep variable names short for variables that are local to the function.
+* Do not export a function or variable name outside the package until you
+  have an external consumer for it.
+* Have setter or getter interfaces/methods to access/manipulate information in
```
Comment (phlogistonjohn) [inline docs/coding.md]: I had to re-read this sentence four or five time before I understood what the recommendation was. Can we word this more clearly?

File: docs/coding.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+We use the following convention for specifying imports:
+
+```
+<import standard library packages>
+
+<import ceph-csi packages>
+
+<import third-party packages>
+```
```
Comment (phlogistonjohn) [inline docs/coding.md]: Is this really the convention here? I thought the proper convention was:
* stdlib
* 3rdparty
* local

If you look at the documenatation for "goimports" for example it says:
```
  -local string
        put imports beginning with this string after 3rd-party packages; comma-separated list
```

File: docs/coding.md (hunk 6 above)
Comment (humblec) [reply docs/coding.md]: Removed it :)

File: docs/coding.md (hunk 7 above)
Comment (humblec) [reply docs/coding.md]: Yes, corrected it.

**Comment by humblec** [comment]
@phlogistonjohn PTAL .. I have addressed the comments.

File: docs/coding.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+Here's a list of some more specific conventions that are often followed in
+the code and will be pointed out in the review process:
+
+## General
+
+* Keep variable names short for variables that are local to the function.
+* Do not export a function or variable name outside the package until you
+  have an external consumer for it.
+* Do not use named return values in function definitions. Use only the type.
```
Comment (ShyamsundarR) [inline docs/coding.md]: @Madhu-1 and myself had a disagreement over this, I am fine with the rule being called out, but I would like to know why we should not use named returns and only use naked returns?
The answer of common style across code does not fit IMO as named returns are for short functions and not for all functions as per the language definition, hence all code cannot accommodate a single style.

File: README.md [hunk 9]
```diff
@@ -20,6 +20,12 @@ CephFS CSI plugins, see documentation in `docs/`.
 
 For example usage of RBD and CephFS CSI plugins, see examples in `examples/`.
 
+## Contributing to this repo
+
+Please follow [contibution guide](https://github.com/ceph/ceph-csi/tree/master/docs/coding.md)
```
Comment (ShyamsundarR) [inline README.md]: I would not call the coding.md a contribution guide, I would call it the coding style guidelines, and hence reorder the links in this statement, with dev-guide first and then coding guidelines next.

File: docs/coding.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+### Error Handling
+
+* Use variable name `err` to denote error variable during a function call.
+* Reuse the previously declared `err` variable as long as it is in scope.
+  For example, do not use `errWrite` or `errRead`.
+* Do not panic() for errors that can be bubbled up back to user. Use panic()
+  only for fatal errors which shouldn't occur.
+* Do not ignore errors using `_` variable unless you know what you're doing.
+* Error strings should not start with a capital letter.
```
Comment (ShyamsundarR) [inline docs/coding.md]: Linter catches this, as a result I would remove this from this list. Less rules to remember for the contributor.

File: docs/coding.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+* If error requires passing of extra information, you can define a new type
+* Error types should end with `Error`.
```
Comment (ShyamsundarR) [inline docs/coding.md]: The above 2 possibly need examples

File: docs/coding.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+### Logging
+
+* If a function is only invoked as part of a transaction step, always use the
```
Comment (ShyamsundarR) [inline docs/coding.md]: What is a transaction? Which ID would we be using here?

**Comment by Madhu-1** [comment]
@humblec anything pending on this one?

**Comment by humblec** [comment]
@Madhu-1 all the comments are addressed. PTAL

**Comment by humblec** [comment]
Thanks @Madhu-1 , merging this.

File: README.md [hunk 13]
```diff
@@ -1,7 +1,7 @@
 # Ceph CSI 1.0.0
 
-[Container Storage Interface
-(CSI)](https://github.com/container-storage-interface/) driver, provisioner,
+This repo containes [Container Storage Interface(CSI)]
```
Comment (Madhu-1) [inline README.md]: containes/ contains

**Pull request merged**